lib/core/src/Cardano/Wallet/Primitive/BlockSummary.hs

```diff
@@ -0,0 +1,200 @@
+        -- INVARIANT: The list is ordered by ascending index.
+    , delegations :: [(Int, DelegationCertificate)]
+        -- ^ (Index of the delegation within the block, delegation data)
+        -- INVARIANT: The list is ordered by ascending index.
+    } deriving (Eq, Ord, Generic, Show)
+
+nullBlockEvents :: BlockEvents -> Bool
+nullBlockEvents BlockEvents{transactions=[],delegations=[]} = True
+nullBlockEvents _ = False
+
+-- | Merge block events that belong to the same block.
+mergeSameBlock :: BlockEvents -> BlockEvents -> BlockEvents
+mergeSameBlock
+    BlockEvents{slot,blockHeight,transactions=txs1,delegations=dlg1}
+    BlockEvents{transactions=txs2,delegations=dlg2}
+  = BlockEvents
+    { slot
+    , blockHeight
+    , transactions = mergeOn fst const txs1 txs2
+    , delegations = mergeOn fst const dlg1 dlg2
+    }
+
+-- | Merge two lists in sorted order. Remove duplicate items.
+mergeOn :: Ord key => (a -> key) -> (a -> a -> a) -> [a] -> [a] -> [a]
+mergeOn _ _ [] ys = ys
+mergeOn _ _ xs [] = xs
+mergeOn f g (x:xs) (y:ys) = case compare (f x) (f y) of
+    LT -> x : mergeOn f g xs (y:ys)
+    EQ -> mergeOn f g (g x y:xs) ys
+    GT -> y : mergeOn f g (x:xs) ys
+
+-- | Get the 'BlockEvents' corresponding to an entire 'Block'.
+fromEntireBlock :: Block -> BlockEvents
+fromEntireBlock Block{header,transactions,delegations} = BlockEvents
+    { slot = toSlot $ chainPointFromBlockHeader header
+    , blockHeight = Block.blockHeight header
+    , transactions = zip [0..] transactions
+    , delegations = zip [0..] delegations
+    }
+
+{-------------------------------------------------------------------------------
+    Testing
+-------------------------------------------------------------------------------}
+-- | For testing:
+-- Convert a list of blocks into a 'BlockSummary'.
+-- Unfortunately, as 'TxIn' references are not resolved,
+-- we can only find transactions with relevant 'TxOut'.
+summarizeOnTxOut :: NonEmpty Block -> LightSummary Identity
+summarizeOnTxOut bs = BlockSummary
+    { from = header . NE.head $ bs
+    , to = header . NE.last $ bs
+    , query = \q -> pure
+        . fromBlockEvents . map (filterBlock q) $ NE.toList bs
+    }
+
+filterBlock :: Either Address RewardAccount -> Block -> BlockEvents
+filterBlock question block = case fromEntireBlock block of
+    BlockEvents{slot,blockHeight,transactions,delegations} -> BlockEvents
+        { slot
+        , blockHeight
+        , transactions = case question of
+            Left addr -> filter (isRelevantTx addr) transactions
+            Right _   -> []
+        , delegations = case question of
+            Left _     -> []
+            Right racc -> filter (isRelevantDelegation racc) delegations
+        }
+  where
+    -- NOTE: Currently used the full address,
+    -- containing both payment and staking parts.
+    -- We may want to query only for the payment part at some point.
+    isRelevantTx addr = any ((addr ==) . address) . outputs . snd
```

Comment:
this (is. ==) will be needed to generalize later as it only survives non-script case. For scripts we will need isShared, here maybe we should sneek isOurs, even at this moment. What do you think?

Reply:
yes, I see it. in light mode we use `query` and it is up to caller  using `BlockSummary` how `query` looks like. So it is upt to caller, ie. consumer of this abstraction, do deal with address management. It is the caller's responsibility

Reply:
Yes, exactly. Hence no `isOurs` here.